# CREPE (CreatePrimersandEvalulate)

## Overview

CREPE is a batch primer design and specificity analysis tool. It uses Primer3 (https://primer3.org/) to create primers from an input CSV of target sites. It then uses UCSC's In-Silico PCR (https://genome.ucsc.edu/cgi-bin/hgPcr) to identify off-target enrichment sites for each primer pair. Lastly, a series of Python functions perform specificity analysis to determine troublesome off-target sites. 

This iteration of CREPE is designed specifically to create primer pairs for 150bp short read sequencing. As such, it first attempts to find primer pairs that generate a 200bp amplicon. We called these 'ideal'. We've found that an 'ideal' primer pair will be created for about 70% of input target sites. For the remaining input sites, CREPE will create a right-anchored primer pair and, if necessary, a left-anchored primer pair. The anchor refers to one of the primers being bound close to the variant of interest while the opposing primer has a larger sequence to attempt to bind to. For example, the right-anchored primer pairs give the forward primer 300bp to find a binding site while the reverse primer will bind close to the target site (within 100bp). The amplicons generated by these anchored pairs can be as large as 500bp. In this way, CREPE attempts to create functional primer pairs for every input target site. 

## Installation

CREPE runs with few dependencies and a simple conda environment.

To begin, download the env.yml file and run the following command:

    conda env create --name p3i --file=env.yml
